cannot convert to 'bool'

First of all, congratulations for this really awesome software. I am currently trying to implement some ODE's solving algorithms with pythran to benchmark the performance and to learn this tool for my research. 
This code is for solving the wave equation in time: 


#pythran export wave(int)

import numpy as np

def wave(n):

    Nx = n
    number_of_threads = 4
    t0 = 0
    x0 = 0
    alpha = np.sqrt(2.)/np.pi
    dx = 12./Nx
    dt = 0.5*dx/alpha
    Nt = int(5.*np.pi/dt)
    c = alpha*dt/dx
    u = np.zeros((Nx,Nt))
    x = np.zeros((Nx,1))
    t = np.zeros((Nt,1))

    #omp parallel for num_threads(number_of_threads) private(i) shared(Nx,x0,x,dx)
    for i in range(Nx):
        x[i] = x0+i*dx

    #omp parallel for num_threads(number_of_threads) private(i) shared(Nt,t0,dt,t)
    for i in range(Nt):
        t[i] = t0 + i*dt
    
    for i in range(Nx):
        if x[i]>=2 and x[i]<=4:
            u[i][0] = np.sin(np.pi*x[i])
        else:
            u[i][0] = 0
        #omp parallel for shared(Nt,u) private(j)
        for j in range(Nt):
            u[0][j]=0
            u[Nx-1][j]=0

    for j in range(Nt-1):
        #omp parallel for num_threads(number_of_threads) private(i) shared(u,Nx,j,c)
        for i in range(Nx-1):
            u[i][j+1]=u[i][j]-(c/2.)*(u[i+1][j]-u[i-1][j])+(np.power(c,2.)/2.)*(u[i+1][j]-2.*u[i][j]+u[i-1][j] )
    return u



The problem occurs in line    if x[i]>2 and x[i]<4:   
The error output is huge which I dont understand fully. 
Compilation command I use: pythran -DUSE_BOOST_SIMD -march=native -Ofast -fopenmp wave.py

The error log: 

/tmp/tmpeser9znt.cpp: In instantiation of ‘typename __pythran_wave::wave::type<argument_type0>::result_
type __pythran_wave::wave::operator()(argument_type0&&) const [with argument_type0 = long int&; typenam
e __pythran_wave::wave::type<argument_type0>::result_type = {anonymous}::pythonic::types::ndarray<doubl
e, 2>]’:                                                                                              
/tmp/tmpeser9znt.cpp:224:68:   required from here
/tmp/tmpeser9znt.cpp:190:11: error: could not convert ‘{anonymous}::pythonic::__builtin__::pythran::and
_(T0&&, T1&&) [with T0 = __pythran_wave::wave::operator()(argument_type0&&) const [with argument_type0 
= long int&; typename __pythran_wave::wave::type<argument_type0>::result_type = {anonymous}::pythonic::
types::ndarray<double, 2>]::<lambda()>; T1 = __pythran_wave::wave::operator()(argument_type0&&) const [
with argument_type0 = long int&; typename __pythran_wave::wave::type<argument_type0>::result_type = {an
onymous}::pythonic::types::ndarray<double, 2>]::<lambda()>; {anonymous}::pythonic::types::lazy_combined
_t<T0, T1> = {anonymous}::pythonic::types::ndarray<bool, 1>](<lambda closure object>__pythran_wave::wav
e::operator()(argument_type0&&) const [with argument_type0 = long int&; typename __pythran_wave::wave::
type<argument_type0>::result_type = {anonymous}::pythonic::types::ndarray<double, 2>]::<lambda()>{x, i}
)’ from ‘{anonymous}::pythonic::types::lazy_combined_t<__pythran_wave::wave::operator()(argument_type0&
&) const [with argument_type0 = long int&; typename __pythran_wave::wave::type<argument_type0>::result_
type = {anonymous}::pythonic::types::ndarray<double, 2>]::<lambda()>, __pythran_wave::wave::operator()(
argument_type0&&) const [with argument_type0 = long int&; typename __pythran_wave::wave::type<argument_
type0>::result_type = {anonymous}::pythonic::types::ndarray<double, 2>]::<lambda()> > {aka {anonymous}:
:pythonic::types::ndarray<bool, 1>}’ to ‘bool’                                                        
           if (pythonic::__builtin__::pythran::and_([&] () { return (x.fast(i) >= 2L); }, [&] () { retu
rn (x.fast(i) <= 4L); }))                                                                             
           ^~
/tmp/tmpeser9znt.cpp:192:36: error: cannot convert ‘std::enable_if<true, {anonymous}::pythonic::types::
numpy_expr<{anonymous}::pythonic::numpy::functor::sin, {anonymous}::pythonic::types::numpy_expr<{anonym
ous}::pythonic::operator_::functor::mul, {anonymous}::pythonic::types::broadcast<double, double>, {anon
ymous}::pythonic::types::numpy_iexpr<const {anonymous}::pythonic::types::ndarray<double, 2>&> > > >::ty
pe {aka {anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::sin, {anonymous
}::pythonic::types::numpy_expr<{anonymous}::pythonic::operator_::functor::mul, {anonymous}::pythonic::t
ypes::broadcast<double, double>, {anonymous}::pythonic::types::numpy_iexpr<const {anonymous}::pythonic:
:types::ndarray<double, 2>&> > >}’ to ‘double’ in assignment                                          
             std::get<0>(u.fast(i)) = pythonic::numpy::functor::sin{}((3.141592653589793 * x.fast(i)));
             ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /tmp/tmpeser9znt.cpp:22:0:
/usr/local/lib/python3.6/site-packages/pythran/pythonic/__builtin__/pythran/and_.hpp: In instantiation 
of ‘{anonymous}::pythonic::types::lazy_combined_t<T0, T1> {anonymous}::pythonic::__builtin__::pythran::
and_(T0&&, T1&&) [with T0 = __pythran_wave::wave::operator()(argument_type0&&) const [with argument_typ
e0 = long int&; typename __pythran_wave::wave::type<argument_type0>::result_type = {anonymous}::pythoni
c::types::ndarray<double, 2>]::<lambda()>; T1 = __pythran_wave::wave::operator()(argument_type0&&) cons
t [with argument_type0 = long int&; typename __pythran_wave::wave::type<argument_type0>::result_type = 
{anonymous}::pythonic::types::ndarray<double, 2>]::<lambda()>; {anonymous}::pythonic::types::lazy_combi
ned_t<T0, T1> = {anonymous}::pythonic::types::ndarray<bool, 1>]’:                                     
/tmp/tmpeser9znt.cpp:190:51:   required from ‘typename __pythran_wave::wave::type<argument_type0>::resu
lt_type __pythran_wave::wave::operator()(argument_type0&&) const [with argument_type0 = long int&; type
name __pythran_wave::wave::type<argument_type0>::result_type = {anonymous}::pythonic::types::ndarray<do
uble, 2>]’                                                                                            
/tmp/tmpeser9znt.cpp:224:68:   required from here
/usr/local/lib/python3.6/site-packages/pythran/pythonic/__builtin__/pythran/and_.hpp:21:7: error: could
 not convert ‘val0’ from ‘{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::operator_::fu
nctor::ge, {anonymous}::pythonic::types::numpy_iexpr<const {anonymous}::pythonic::types::ndarray<double
, 2>&>, {anonymous}::pythonic::types::broadcast<double, long int> >’ to ‘bool’                        
       if (val0)
       ^~
/tmp/tmpeser9znt.cpp: In instantiation of ‘typename __pythran_wave::wave::type<argument_type0>::result_
type __pythran_wave::wave::operator()(argument_type0&&) const [with argument_type0 = long int&; typenam
e __pythran_wave::wave::type<argument_type0>::result_type = {anonymous}::pythonic::types::ndarray<doubl
e, 2>]’:                                                                                              
/tmp/tmpeser9znt.cpp:224:68:   required from here
/tmp/tmpeser9znt.cpp:190:11: error: could not convert ‘{anonymous}::pythonic::__builtin__::pythran::and
_(T0&&, T1&&) [with T0 = __pythran_wave::wave::operator()(argument_type0&&) const [with argument_type0 
= long int&; typename __pythran_wave::wave::type<argument_type0>::result_type = {anonymous}::pythonic::
types::ndarray<double, 2>]::<lambda()>; T1 = __pythran_wave::wave::operator()(argument_type0&&) const [
with argument_type0 = long int&; typename __pythran_wave::wave::type<argument_type0>::result_type = {an
onymous}::pythonic::types::ndarray<double, 2>]::<lambda()>; {anonymous}::pythonic::types::lazy_combined
_t<T0, T1> = {anonymous}::pythonic::types::ndarray<bool, 1>](<lambda closure object>__pythran_wave::wav
e::operator()(argument_type0&&) const [with argument_type0 = long int&; typename __pythran_wave::wave::
type<argument_type0>::result_type = {anonymous}::pythonic::types::ndarray<double, 2>]::<lambda()>{x, i}
)’ from ‘{anonymous}::pythonic::types::lazy_combined_t<__pythran_wave::wave::operator()(argument_type0&
&) const [with argument_type0 = long int&; typename __pythran_wave::wave::type<argument_type0>::result_
type = {anonymous}::pythonic::types::ndarray<double, 2>]::<lambda()>, __pythran_wave::wave::operator()(
argument_type0&&) const [with argument_type0 = long int&; typename __pythran_wave::wave::type<argument_
type0>::result_type = {anonymous}::pythonic::types::ndarray<double, 2>]::<lambda()> > {aka {anonymous}:
:pythonic::types::ndarray<bool, 1>}’ to ‘bool’                                                        
           if (pythonic::__builtin__::pythran::and_([&] () { return (x.fast(i) >= 2L); }, [&] () { retu
rn (x.fast(i) <= 4L); }))                                                                             
           ^~
/tmp/tmpeser9znt.cpp:192:36: error: cannot convert ‘std::enable_if<true, {anonymous}::pythonic::types::
numpy_expr<{anonymous}::pythonic::numpy::functor::sin, {anonymous}::pythonic::types::numpy_expr<{anonym
ous}::pythonic::operator_::functor::mul, {anonymous}::pythonic::types::broadcast<double, double>, {anon
ymous}::pythonic::types::numpy_iexpr<const {anonymous}::pythonic::types::ndarray<double, 2>&> > > >::ty
pe {aka {anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::sin, {anonymous
}::pythonic::types::numpy_expr<{anonymous}::pythonic::operator_::functor::mul, {anonymous}::pythonic::t
ypes::broadcast<double, double>, {anonymous}::pythonic::types::numpy_iexpr<const {anonymous}::pythonic:
:types::ndarray<double, 2>&> > >}’ to ‘double’ in assignment                                          
             std::get<0>(u.fast(i)) = pythonic::numpy::functor::sin{}((3.141592653589793 * x.fast(i)));
             ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /tmp/tmpeser9znt.cpp:22:0:
/usr/local/lib/python3.6/site-packages/pythran/pythonic/__builtin__/pythran/and_.hpp: In instantiation 
of ‘{anonymous}::pythonic::types::lazy_combined_t<T0, T1> {anonymous}::pythonic::__builtin__::pythran::
and_(T0&&, T1&&) [with T0 = __pythran_wave::wave::operator()(argument_type0&&) const [with argument_typ
e0 = long int&; typename __pythran_wave::wave::type<argument_type0>::result_type = {anonymous}::pythoni
c::types::ndarray<double, 2>]::<lambda()>; T1 = __pythran_wave::wave::operator()(argument_type0&&) cons
t [with argument_type0 = long int&; typename __pythran_wave::wave::type<argument_type0>::result_type = 
{anonymous}::pythonic::types::ndarray<double, 2>]::<lambda()>; {anonymous}::pythonic::types::lazy_combi
ned_t<T0, T1> = {anonymous}::pythonic::types::ndarray<bool, 1>]’:                                     
/tmp/tmpeser9znt.cpp:190:51:   required from ‘typename __pythran_wave::wave::type<argument_type0>::resu
lt_type __pythran_wave::wave::operator()(argument_type0&&) const [with argument_type0 = long int&; type
name __pythran_wave::wave::type<argument_type0>::result_type = {anonymous}::pythonic::types::ndarray<do
uble, 2>]’                                                                                            
/tmp/tmpeser9znt.cpp:224:68:   required from here
/usr/local/lib/python3.6/site-packages/pythran/pythonic/__builtin__/pythran/and_.hpp:21:7: error: could
 not convert ‘val0’ from ‘{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::operator_::fu
nctor::ge, {anonymous}::pythonic::types::numpy_iexpr<const {anonymous}::pythonic::types::ndarray<double
, 2>&>, {anonymous}::pythonic::types::broadcast<double, long int> >’ to ‘bool’                        
       if (val0)
       ^~
WARNING  Compilation error, trying hard to find its origin...
WARNING  Nop, I'm going to flood you with C++ errors!
CRITICAL Cover me Jack. Jack? Jaaaaack!!!!
E: error: Command "g++ -pthread -Wno-unused-result -Wsign-compare -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDE
BUG -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-s
trong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m
64 -mtune=generic -fasynchronous-unwind-tables -D_GNU_SOURCE -fPIC -fwrapv -fPIC -DENABLE_PYTHON_MODULE
 -D__PYTHRAN__=3 -DUSE_BOOST_SIMD -DPYTHRAN_BLAS_BLAS -I/usr/local/lib/python3.6/site-packages/pythran 
-I/usr/local/lib64/python3.6/site-packages/numpy/core/include -I/usr/local/lib64/python3.6/site-package
s/numpy/core/include -I/usr/include/python3.6m -c /tmp/tmpeser9znt.cpp -o /tmp/tmpz_3lvxpu/tmp/tmpeser9
znt.o -std=c++11 -fno-math-errno -w -fvisibility=hidden -fno-wrapv -march=native -Ofast -fopenmp" faile
d with exit status 1 

I searched all over your documentation and didnt find any solution to this. Could you please help me out? It is critical if I will use pythran to write scientific computations if I can have logical statements with arrays like x[i]>=some number. 

thank you in advance

Alright, I just cloned the repo and installed it. And the 'bool' error is gone! But...


/tmp/tmpejhhh0pg.cpp: In instantiation of ‘typename __pythran_wave::wave::type<argument_type0>::result_
type __pythran_wave::wave::operator()(argument_type0&&) const [with argument_type0 = long int&; typenam
e __pythran_wave::wave::type<argument_type0>::result_type = {anonymous}::pythonic::types::ndarray<doubl
e, 2>]’:                                                                                              
/tmp/tmpejhhh0pg.cpp:201:68:   required from here
/tmp/tmpejhhh0pg.cpp:181:34: error: cannot convert ‘std::enable_if<true, {anonymous}::pythonic::types::
numpy_expr<{anonymous}::pythonic::numpy::functor::sin, {anonymous}::pythonic::types::numpy_expr<{anonym
ous}::pythonic::operator_::functor::mul, {anonymous}::pythonic::types::broadcast<double, double>, {anon
ymous}::pythonic::types::numpy_iexpr<const {anonymous}::pythonic::types::ndarray<double, 2>&> > > >::ty
pe {aka {anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::sin, {anonymous
}::pythonic::types::numpy_expr<{anonymous}::pythonic::operator_::functor::mul, {anonymous}::pythonic::t
ypes::broadcast<double, double>, {anonymous}::pythonic::types::numpy_iexpr<const {anonymous}::pythonic:
:types::ndarray<double, 2>&> > >}’ to ‘double’ in assignment                                          
           std::get<0>(u.fast(i)) = pythonic::numpy::functor::sin{}((3.141592653589793 * x.fast(i)));
           ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/tmpejhhh0pg.cpp: In instantiation of ‘typename __pythran_wave::wave::type<argument_type0>::result_
type __pythran_wave::wave::operator()(argument_type0&&) const [with argument_type0 = long int&; typenam
e __pythran_wave::wave::type<argument_type0>::result_type = {anonymous}::pythonic::types::ndarray<doubl
e, 2>]’:                                                                                              
/tmp/tmpejhhh0pg.cpp:201:68:   required from here
/tmp/tmpejhhh0pg.cpp:181:34: error: cannot convert ‘std::enable_if<true, {anonymous}::pythonic::types::
numpy_expr<{anonymous}::pythonic::numpy::functor::sin, {anonymous}::pythonic::types::numpy_expr<{anonym
ous}::pythonic::operator_::functor::mul, {anonymous}::pythonic::types::broadcast<double, double>, {anon
ymous}::pythonic::types::numpy_iexpr<const {anonymous}::pythonic::types::ndarray<double, 2>&> > > >::ty
pe {aka {anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::sin, {anonymous
}::pythonic::types::numpy_expr<{anonymous}::pythonic::operator_::functor::mul, {anonymous}::pythonic::t
ypes::broadcast<double, double>, {anonymous}::pythonic::types::numpy_iexpr<const {anonymous}::pythonic:
:types::ndarray<double, 2>&> > >}’ to ‘double’ in assignment                                          
           std::get<0>(u.fast(i)) = pythonic::numpy::functor::sin{}((3.141592653589793 * x.fast(i)));
           ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
WARNING  Compilation error, trying hard to find its origin...
WARNING  Nop, I'm going to flood you with C++ errors!
CRITICAL Cover me Jack. Jack? Jaaaaack!!!!
E: error: Command "g++ -pthread -Wno-unused-result -Wsign-compare -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDE
BUG -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-s
trong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m
64 -mtune=generic -fasynchronous-unwind-tables -D_GNU_SOURCE -fPIC -fwrapv -fPIC -DENABLE_PYTHON_MODULE
 -D__PYTHRAN__=3 -DUSE_BOOST_SIMD -DPYTHRAN_BLAS_BLAS -I/usr/local/lib/python3.6/site-packages/pythran-
0.8.5-py3.6.egg/pythran -I/usr/local/lib64/python3.6/site-packages/numpy/core/include -I/usr/local/lib6
4/python3.6/site-packages/numpy/core/include -I/usr/include/python3.6m -c /tmp/tmpejhhh0pg.cpp -o /tmp/
tmpz6s6knor/tmp/tmpejhhh0pg.o -std=c++11 -fno-math-errno -w -fvisibility=hidden -fno-wrapv -march=nativ
e -Ofast -fopenmp" failed with exit status 1   

 
this came up this time... It has to do with the line: u[i][0] = np.sin(np.pi*x[i])
is it my fault or because of the new feature ndarray.bool ?

Changing the above line to u[i][0] = x[i] ( because the u[i][0] = np.sin(np.pi *i) for example worked )
i get this message if it helps more:

/tmp/tmph9_pgsdd.cpp: In instantiation of ‘typename __pythran_wave::wave::type<argument_type0>::result_
type __pythran_wave::wave::operator()(argument_type0&&) const [with argument_type0 = long int&; typenam
e __pythran_wave::wave::type<argument_type0>::result_type = {anonymous}::pythonic::types::ndarray<doubl
e, 2>]’:                                                                                              
/tmp/tmph9_pgsdd.cpp:193:68:   required from here
/tmp/tmph9_pgsdd.cpp:173:34: error: cannot convert ‘{anonymous}::pythonic::types::numpy_iexpr<const {an
onymous}::pythonic::types::ndarray<double, 2>&>’ to ‘double’ in assignment                            
           std::get<0>(u.fast(i)) = x.fast(i);
/tmp/tmph9_pgsdd.cpp: In instantiation of ‘typename __pythran_wave::wave::type<argument_type0>::result_
type __pythran_wave::wave::operator()(argument_type0&&) const [with argument_type0 = long int&; typenam
e __pythran_wave::wave::type<argument_type0>::result_type = {anonymous}::pythonic::types::ndarray<doubl
e, 2>]’:                                                                                              
/tmp/tmph9_pgsdd.cpp:193:68:   required from here
/tmp/tmph9_pgsdd.cpp:173:34: error: cannot convert ‘{anonymous}::pythonic::types::numpy_iexpr<const {an
onymous}::pythonic::types::ndarray<double, 2>&>’ to ‘double’ in assignment                            
           std::get<0>(u.fast(i)) = x.fast(i);
WARNING  Compilation error, trying hard to find its origin...
WARNING  Nop, I'm going to flood you with C++ errors!
CRITICAL Cover me Jack. Jack? Jaaaaack!!!!
E: error: Command "g++ -pthread -Wno-unused-result -Wsign-compare -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDE
BUG -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-s
trong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m
64 -mtune=generic -fasynchronous-unwind-tables -D_GNU_SOURCE -fPIC -fwrapv -fPIC -DENABLE_PYTHON_MODULE
 -D__PYTHRAN__=3 -DUSE_BOOST_SIMD -DPYTHRAN_BLAS_BLAS -I/usr/local/lib/python3.6/site-packages/pythran-
0.8.5-py3.6.egg/pythran -I/usr/local/lib64/python3.6/site-packages/numpy/core/include -I/usr/local/lib6
4/python3.6/site-packages/numpy/core/include -I/usr/include/python3.6m -c /tmp/tmph9_pgsdd.cpp -o /tmp/
tmpe8svh3za/tmp/tmph9_pgsdd.o -std=c++11 -fno-math-errno -w -fvisibility=hidden -fno-wrapv -march=nativ
e -Ofast -fopenmp" failed with exit status 1

--- a/pythran/pythonic/include/types/ndarray.hpp
+++ b/pythran/pythonic/include/types/ndarray.hpp
@@ -468,6 +468,8 @@ namespace types
         ndarray<T, M> reshape(array<long, M> const &shape) &&
         ;
 
+    explicit operator bool() const;
+
     ndarray<T, 1> flat() const;
     ndarray<T, N> copy() const;
     intptr_t id() const;

--- a/pythran/pythonic/include/types/numpy_broadcast.hpp
+++ b/pythran/pythonic/include/types/numpy_broadcast.hpp
@@ -242,6 +242,8 @@ namespace types
     broadcast(V v);
 
     dtype operator[](long) const;
+    template <size_t N>
+    dtype operator[](array<long, N>) const;
     broadcast operator[](slice) const
     {
       return *this;

--- a/pythran/pythonic/include/types/numpy_expr.hpp
+++ b/pythran/pythonic/include/types/numpy_expr.hpp
@@ -603,6 +603,13 @@ namespace types
       return _index(s, utils::make_index_sequence<sizeof...(Args)>{});
     }
 
+    dtype operator[](array<long, value> const &indices) const
+    {
+      return _index(indices, utils::make_index_sequence<sizeof...(Args)>{});
+    }
+
+    explicit operator bool() const;
+
     long flat_size() const;
 
     long size() const;

--- a/pythran/pythonic/include/types/numpy_gexpr.hpp
+++ b/pythran/pythonic/include/types/numpy_gexpr.hpp
@@ -593,6 +593,8 @@ namespace types
     auto operator[](long i) const -> decltype(this->fast(i));
     auto operator[](long i) -> decltype(this->fast(i));
 
+    explicit operator bool() const;
+
     long flat_size() const;
     long size() const;
     ndarray<dtype, value> copy() const

--- a/pythran/pythonic/include/types/numpy_iexpr.hpp
+++ b/pythran/pythonic/include/types/numpy_iexpr.hpp
@@ -195,6 +195,7 @@ namespace types
       return nget<M - 1>()(*this, indices);
     }
 
+    explicit operator bool() const;
     long flat_size() const;
     array<long, value> const &shape() const
     {

--- a/pythran/pythonic/include/types/numpy_texpr.hpp
+++ b/pythran/pythonic/include/types/numpy_texpr.hpp
@@ -157,6 +157,7 @@ namespace types
             std::tuple<S0 const &, S const &...>{s0, s...},
             utils::make_reversed_index_sequence<1 + sizeof...(S)>()));
 
+    explicit operator bool() const;
     long flat_size() const;
     intptr_t id() const;
     ndarray<dtype, value> copy() const

--- a/pythran/pythonic/types/ndarray.hpp
+++ b/pythran/pythonic/types/ndarray.hpp
@@ -7,6 +7,8 @@
 #include "pythonic/types/empty_iterator.hpp"
 #include "pythonic/types/attr.hpp"
 
+#include "pythonic/__builtin__/ValueError.hpp"
+
 #include "pythonic/utils/nested_container.hpp"
 #include "pythonic/utils/shared_ref.hpp"
 #include "pythonic/utils/reserve.hpp"
@@ -769,6 +771,16 @@ namespace types
   ndarray<T, M> ndarray<T, N>::reshape(array<long, M> const &shape) &&
   {
     return {std::move(mem), shape};
+  }
+
+  template <class T, size_t N>
+  ndarray<T, N>::operator bool() const
+  {
+    if (std::any_of(_shape.begin(), _shape.end(),
+                    [](long n) { return n != 1; }))
+      throw ValueError("The truth value of an array with more than one element "
+                       "is ambiguous. Use a.any() or a.all()");
+    return *buffer;
   }
 
   template <class T, size_t N>

--- a/pythran/pythonic/types/numpy_broadcast.hpp
+++ b/pythran/pythonic/types/numpy_broadcast.hpp
@@ -98,10 +98,19 @@ namespace types
   }
 
   template <class T, class B>
+  template <size_t N>
+  typename broadcast<T, B>::dtype broadcast<T, B>::
+  operator[](array<long, N>) const
+  {
+    return _base._value;
+  }
+
+  template <class T, class B>
   typename broadcast<T, B>::dtype broadcast<T, B>::fast(long) const
   {
     return _base._value;
   }
+
   template <class T, class B>
   template <class... Args>
   typename broadcast<T, B>::dtype broadcast<T, B>::

--- a/pythran/pythonic/types/numpy_expr.hpp
+++ b/pythran/pythonic/types/numpy_expr.hpp
@@ -6,6 +6,8 @@
 #include "pythonic/utils/meta.hpp"
 #include "pythonic/types/nditerator.hpp"
 
+#include "pythonic/__builtin__/ValueError.hpp"
+
 PYTHONIC_NS_BEGIN
 
 namespace types
@@ -328,6 +330,17 @@ namespace types
       operator[](F const &filter) const
   {
     return fast(filter);
+  }
+
+  template <class Op, class... Args>
+  numpy_expr<Op, Args...>::operator bool() const
+  {
+    if (std::any_of(_shape.begin(), _shape.end(),
+                    [](long n) { return n != 1; }))
+      throw ValueError("The truth value of an array with more than one element "
+                       "is ambiguous. Use a.any() or a.all()");
+    array<long, value> first = {0};
+    return operator[](first);
   }
 
   template <class Op, class... Args>

--- a/pythran/pythonic/types/numpy_gexpr.hpp
+++ b/pythran/pythonic/types/numpy_gexpr.hpp
@@ -3,6 +3,8 @@
 
 #include "pythonic/include/types/numpy_gexpr.hpp"
 
+#include "pythonic/__builtin__/ValueError.hpp"
+
 #include "pythonic/utils/meta.hpp"
 #include "pythonic/operator_/iadd.hpp"
 #include "pythonic/operator_/isub.hpp"
@@ -802,6 +804,16 @@ namespace types
       operator[](F const &filter) const
   {
     return fast(filter);
+  }
+
+  template <class Arg, class... S>
+  numpy_gexpr<Arg, S...>::operator bool() const
+  {
+    if (std::any_of(_shape.begin(), _shape.end(),
+                    [](long n) { return n != 1; }))
+      throw ValueError("The truth value of an array with more than one element "
+                       "is ambiguous. Use a.any() or a.all()");
+    return *buffer;
   }
 
   template <class Arg, class... S>

--- a/pythran/pythonic/types/numpy_iexpr.hpp
+++ b/pythran/pythonic/types/numpy_iexpr.hpp
@@ -374,9 +374,11 @@ namespace types
                         ? indices[value - 1] + _shape[value - 1]
                         : indices[value - 1];
     long mult = _shape[value - 1];
-    for (size_t i = value - 2; i > 0; --i) {
-      offset += (indices[i] < 0 ? indices[i] + _shape[i] : indices[i]) * mult;
-      mult *= _shape[i];
+    if (value > 1) {
+      for (size_t i = value - 2; i > 0; --i) {
+        offset += (indices[i] < 0 ? indices[i] + _shape[i] : indices[i]) * mult;
+        mult *= _shape[i];
+      }
     }
     return buffer[offset +
                   (indices[0] < 0 ? indices[0] + _shape[0] : indices[0]) *
@@ -388,6 +390,16 @@ namespace types
   operator[](array<long, value> const &indices)
   {
     return const_cast<dtype &>(const_cast<numpy_iexpr const &>(*this)[indices]);
+  }
+
+  template <class Arg>
+  numpy_iexpr<Arg>::operator bool() const
+  {
+    if (std::any_of(_shape.begin(), _shape.end(),
+                    [](long n) { return n != 1; }))
+      throw ValueError("The truth value of an array with more than one element "
+                       "is ambiguous. Use a.any() or a.all()");
+    return *buffer;
   }
 
   template <class Arg>

--- a/pythran/pythonic/types/numpy_texpr.hpp
+++ b/pythran/pythonic/types/numpy_texpr.hpp
@@ -248,6 +248,12 @@ namespace types
   }
 
   template <class E>
+  numpy_texpr_2<E>::operator bool() const
+  {
+    return (bool)arg;
+  }
+
+  template <class E>
   long numpy_texpr_2<E>::flat_size() const
   {
     return arg.flat_size();

--- a/pythran/tests/test_ndarray.py
+++ b/pythran/tests/test_ndarray.py
@@ -757,3 +757,34 @@ def assign_ndarray(t):
                       10, 10,
                       numpy_expr_combiner=[int, int])
 
+    def test_bool_conversion_0(self):
+        code = 'def bool_conversion_0(x): return bool(x)'
+        self.run_test(code,
+                      numpy.array([[1]]),
+                      bool_conversion_0=[NDArray[int, :,:]])
+
+    def test_bool_conversion_1(self):
+        code = 'def bool_conversion_1(x): return bool(x > 1)'
+        self.run_test(code,
+                      numpy.array([1]),
+                      bool_conversion_1=[NDArray[int, :]])
+
+    def test_bool_conversion_2(self):
+        code = 'def bool_conversion_2(x): return bool(x[1:,1,1:])'
+        self.run_test(code,
+                      numpy.array([[[1,1],[1,1]],
+                                   [[1,0],[1,1]]]
+                                   ),
+                                   bool_conversion_2=[NDArray[int, :,:,:]])
+
+    def test_bool_conversion_3(self):
+        code = 'def bool_conversion_3(x): return bool(x[1])'
+        self.run_test(code,
+                      numpy.array([[1],[0]]),
+                      bool_conversion_3=[NDArray[int, :,:]])
+
+    def test_bool_conversion_4(self):
+        code = 'def bool_conversion_4(x): return bool(x.T)'
+        self.run_test(code,
+                      numpy.array([[1]]),
+                      bool_conversion_4=[NDArray[int, :,:]])

